filebeat/crawler/registrar.go

```diff
@@ -209,3 +210,22 @@ func (r *Registrar) getPreviousFile(newFilePath string, newFileInfo os.FileInfo)
 
 	return "", fmt.Errorf("No previous file found")
 }
+
+func (r *Registrar) setState(path string, state *FileState) {
+	r.stateMutex.Lock()
+	defer r.stateMutex.Unlock()
+	r.State[path] = state
+}
+
+func (r *Registrar) getState(path string) (*FileState, bool) {
+	r.stateMutex.Lock()
+	defer r.stateMutex.Unlock()
+	state, exist := r.State[path]
+	return state, exist
+}
+
+func (r *Registrar) getFullState() map[string]*FileState {
+	r.stateMutex.Lock()
+	defer r.stateMutex.Unlock()
+	return r.State
```

Comment:
The lock is not providing protection here. You would need to make a copy of `r.State` while holding the lock then return the copy. This assumes that the `FileState` objects are treated as immutable (if not then you need to make a deep copy).